### ISSUE TITLE
Use of pre-existing AWS resources

### ISSUE THREAD

**Issue by richford** [richford/cloudknot#116]
Continuing from discussion on #111. From @dtenenba:

> Basically my questions have to do with adapting cloudknot to run in an AWS account that is shared by many IAM users, where end users do not have full Batch permissions. Specifically, end users cannot create compute environments or queues. They also cannot create S3 buckets, but every user has a bucket (that of their lab or administrative division) they can read/write to/from.
Is it possible to specify a pre-existing compute environment and queue? This would also avoid the problem of running out of compute environments and queues (there are by default pretty low limits on the number of each of these that one account can have).

**Comment by richford** [comment]
Most of the objects in the aws submodule will retrieve a pre-existing resource if you supply an already existing arn or name argument. `ComputeEnvironment` and `JobQueue` both have this capability. And you can specify this at the higher level when creating a `Knot` by using the `compute_environment_name` and `job_queue_name` arguments. Users can specify a pre-existing bucket using the `set_s3_bucket()` function.

**Comment by dtenenba** [comment]
When using these predefined resources, will `clobber` get rid of them? I do not want that to happen (and in any case for end users the result would be a permission denied error).

**Comment by richford** [comment]
Yes, `clobber` will attempt to remove the resources from AWS. I'd like to keep the ability to remove resources that were previously defined. One could imagine a user creating a compute environment, using is for a few weeks, retrieving the pre-existing resource for each new session, and then clobbering that resource at the end of that period.

Once solution for your case is to catch any permission denied errors and turn them into warnings. You would still be responsible for managing the user permissions so that they didn't delete communal resources. But then they could clobber without raising exceptions. How does that approach sound?

**Comment by dtenenba** [comment]
It sounds ok.

I'm wondering though if you could keep track of which resources were supplied to `Knot()` (in other words, which resources already existed before cloudknot came along), and to make the default behavior to never clobber those resources?

That seems a bit more elegant to me....

**Comment by dtenenba** [comment]
FYI, I tried to use my pre-existing bucket and got an error that the bucket already exists. 
Is it possible to have cloudknot use an existing bucket? Our users don't have permission to create buckets (but they do all have a bucket they can write to).

Also (I'm anticipating the next problem) all buckets in our account have a bucket policy attached that requires files be uploaded with server side encryption set to AES256. Is there a way to specify that be used for all cloudknot uploads?

Thanks.

```
import cloudknot as ck
 ck.set_s3_bucket("fh-div-adm-scicomp")
---------------------------------------------------------------------------
BucketAlreadyExists                       Traceback (most recent call last)
~/envs/batch-demo/lib/python3.6/site-packages/cloudknot/aws/base_classes.py in set_s3_bucket(bucket)
    171                 CreateBucketConfiguration={
--> 172                     'LocationConstraint': get_region()
    173                 }

~/envs/batch-demo/lib/python3.6/site-packages/botocore/client.py in _api_call(self, *args, **kwargs)
    313             # The "self" in this scope is referring to the BaseClient.
--> 314             return self._make_api_call(operation_name, kwargs)
    315

~/envs/batch-demo/lib/python3.6/site-packages/botocore/client.py in _make_api_call(self, operation_name, api_params)
    611             error_class = self.exceptions.from_code(error_code)
--> 612             raise error_class(parsed_response, operation_name)
    613         else:

BucketAlreadyExists: An error occurred (BucketAlreadyExists) when calling the CreateBucket operation: The requested bucket name is not available. The bucket namespace is shared by all users of the system. Please select a different name and try again.

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-64-5e40eea6a912> in <module>()
----> 1 ck.set_s3_bucket("fh-div-adm-scicomp")

~/envs/batch-demo/lib/python3.6/site-packages/cloudknot/aws/base_classes.py in set_s3_bucket(bucket)
    176             pass
    177         except clients['s3'].exceptions.BucketAlreadyExists:
--> 178             raise ValueError('The requested bucket name is not available.')
    179         except clients['s3'].exceptions.ClientError as e:
    180             # Check for Illegal Location Constraint

ValueError: The requested bucket name is not available.
```

**Comment by richford** [comment]
Bucket fix in progress, see #121.

**Comment by richford** [comment]
@dtenanba, I believe #121 resolves this issue. You can now use `ck.get_s3_params()`, which returns a named tuple with bucket and policy fields. And you can use `ck.set_s3_params(bucket[, policy])` to specify an s3 bucket. If you don't supply a policy with which to access that bucket, cloudknot will try to create one for you (probably not what you want given your users' permissions). Otherwise, you can specify a policy.

I moved the follow-up issue about SSE keys to #122.

**Comment by richford** [comment]
Closing after #121.
